sgx-tools/get-ias-report.go

```diff
@@ -0,0 +1,151 @@
+package main // import "github.com/inclavare-containers/sgx-tools"
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/go-restruct/restruct"
+	"github.com/opencontainers/runc/libenclave/attestation"
+	"github.com/opencontainers/runc/libenclave/attestation/sgx"
+	"github.com/opencontainers/runc/libenclave/intelsgx"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"io"
+	"log"
+	"os"
+)
+
+var getIasReportCommand = cli.Command{
+	Name:  "get-ias-report",
+	Usage: "get remote attestation report from IAS",
+	ArgsUsage: `[command options]
+
+EXAMPLE:
+For example, get remote attestation report from IAS according to the given local report file:
+
+	# sgx-tools get-ias-report --report foo.rep --spid ${SPID} --subscription-key ${SUBSCRIPTION_KEY}`,
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "report",
+			Usage: "path to the input report file containing REPORT",
+		},
+		cli.BoolFlag{
+			Name:  "product",
+			Usage: "specify whether using production attestation service",
+		},
+		cli.StringFlag{
+			Name:  "spid",
+			Usage: "spid",
+		},
+		cli.StringFlag{
+			Name:  "subscription-key, -key",
+			Usage: "specify the subscription key",
+		},
+		cli.BoolFlag{
+			Name:  "linkable",
+			Usage: "specify the EPID signatures policy type",
+		},
+	},
+	Action: func(context *cli.Context) error {
+		reportPath := context.String("report")
+		if reportPath == "" {
+			return fmt.Errorf("report argument cannot be empty")
+		}
+
+		spid := context.String("spid")
+		if spid == "" {
+			return fmt.Errorf("spid argument cannot be empty")
+		}
+
+		subscriptionKey := context.String("subscription-key")
+		if subscriptionKey == "" {
+			return fmt.Errorf("subscription-key argument cannot be empty")
+		}
+
+		if context.GlobalBool("verbose") {
+			logrus.SetLevel(logrus.DebugLevel)
+		}
+
+		rf, err := os.Open(reportPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return fmt.Errorf("report file %s not found", reportPath)
+			}
+			return err
+		}
+		defer rf.Close()
+
+		var rfi os.FileInfo
+		rfi, err = rf.Stat()
+		if err != nil {
+			return err
+		}
+
+		if rfi.Size() != intelsgx.ReportLength {
+			return fmt.Errorf("report file %s not match REPORT", reportPath)
+		}
+
+		buf := make([]byte, intelsgx.ReportLength)
+		if _, err = io.ReadFull(rf, buf); err != nil {
+			return fmt.Errorf("report file %s read failed", reportPath)
+		}
+
+		linkable := false
+		if context.Bool("linkable") {
+			linkable = true
+		}
+
+		quote, err := intelsgx.GetQuote(buf, spid, linkable)
+		if err != nil {
+			return err
+		}
```

Comment:
Basically above codes are duplicated with gen-quote. Could you move them into a common code file?